packages/blockchain-api/src/currencyConversion/CurrencyConversionAPI.ts

```diff
@@ -0,0 +1,102 @@
+import { DataSource, DataSourceConfig } from 'apollo-datasource'
+import BigNumber from 'bignumber.js'
+import { CurrencyConversionArgs } from '../schema'
+import { CGLD, CUSD, USD } from './consts'
+import ExchangeRateAPI from './ExchangeRateAPI'
+import GoldExchangeRateAPI from './GoldExchangeRateAPI'
+
+function insertIf<T>(condition: boolean, element: T) {
+  return condition ? [element] : []
+}
+
+export default class CurrencyConversionAPI<TContext = any> extends DataSource {
+  // httpCache!: HTTPCache;
+  // context!: TContext;
+  // memoizedResults = new Map<string, Promise<any>>();
+
+  exchangeRateAPI = new ExchangeRateAPI()
+  goldExchangeRateAPI = new GoldExchangeRateAPI()
+
+  initialize(config: DataSourceConfig<TContext>): void {
+    // this.context = config.context;
+    // this.httpCache = new HTTPCache(config.cache);
+    this.exchangeRateAPI.initialize(config)
+    this.goldExchangeRateAPI.initialize(config)
+  }
+
+  async getExchangeRate({
+    sourceCurrencyCode,
+    currencyCode,
+    timestamp,
+  }: CurrencyConversionArgs): Promise<BigNumber> {
+    const fromCode = sourceCurrencyCode || USD
+    const toCode = currencyCode
+
+    let steps = this.getConversionSteps(fromCode, toCode)
+
+    const ratesPromises = []
+    for (let i = 1; i < steps.length; i++) {
+      const prevCode = steps[i - 1]
+      const code = steps[i]
+      ratesPromises.push(this.getSupportedExchangeRate(prevCode, code, timestamp))
+    }
+
+    const rates = await Promise.all(ratesPromises)
+
+    // Multiply all rates
+    const rate = rates.reduce((acc, rate) => acc.multipliedBy(rate), new BigNumber(1))
+
+    return rate
+  }
+
+  private getConversionSteps(fromCode: string, toCode: string) {
+    if (fromCode === toCode) {
+      // Same code, nothing to do
+      return []
+    } else if (fromCode === CGLD || toCode === CGLD) {
+      // cGLD -> X (where X !== cUSD)
+      if (fromCode === CGLD && toCode !== CUSD) {
+        return [CGLD, CUSD, ...insertIf(toCode !== USD, USD), toCode]
```

Comment:
Do we assume that the cUSD -> local currency rate is the same as the cUSD -> USD -> local currency rate? This may not be the case if the price of cUSD differs across markets. For example, there used to be a consistent spread between BTC -> dollars and BTC -> Korean yen

One workaround, @nityas suggested in the cash in/out spec is to use cUSD -> local currency rate as the stored exchange rate. Not sure whether it's easier to address now or during some of the cash in/out work

Reply:
Yes that's the assumption so far given the data we have.
I imagined this would have to change as soon as we have data with more direct markets.

This function is the generic way to get exchange rates but the idea of taking the rate computed from exchanges seems good to me and is something I thought about when designing the new blockchain-api schema changes. I had however only thought about this in the context of exchange transactions for a specific address.
From what you wrote, it looks like @nityas thought we could store exchanges rates for all exchange transactions as they happen? this seems very appealing, would love to chat directly about this!